framework/tlvf/src/src/WSC/AttrList.cpp

```diff
@@ -0,0 +1,504 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * Copyright (c) 2019 Tomer Eliyahu (Intel)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <tlvf/WSC/AttrList.h>
+#include <tlvf/tlvflogging.h>
+
+using namespace WSC;
+
+std::shared_ptr<AttrList> AttrList::create(ieee1905_1::tlvWsc &tlv, const config &cfg)
+{
+    std::shared_ptr<AttrList> attributes = nullptr;
+    if (cfg.msg_type == eWscMessageType::WSC_MSG_TYPE_M1) {
+        attributes = std::make_shared<m1>(tlv.payload(), tlv.payload_length(), false);
+        if (!attributes->init(cfg))
+            return nullptr;
+        attributes->finalize();
+        tlv.addInnerClassList(attributes);
+    } else if (cfg.msg_type == eWscMessageType::WSC_MSG_TYPE_M2) {
+        attributes = std::make_shared<m2>(tlv.payload(), tlv.payload_length(), false);
+        if (!attributes->init(cfg))
+            return nullptr;
+        attributes->finalize();
+        tlv.addInnerClassList(attributes);
+    }
+
+    return attributes;
+}
+
+eWscMessageType AttrList::wsc_msg_type_from_buffer(uint8_t *buff, size_t buff_len)
+{
+    auto attr                = reinterpret_cast<sWscAttrHeader *>(buff);
+    size_t attributes_length = 0;
+    while (ntohs(attr->type) != ATTR_MSG_TYPE) {
+        attributes_length += ntohs(attr->length) + sizeof(sWscAttrHeader);
+        if (attributes_length > buff_len)
+            return eWscMessageType::WSC_MSG_TYPE_INVALID;
+        attr = reinterpret_cast<sWscAttrHeader *>(buff + attributes_length);
+    }
+    return static_cast<eWscMessageType>(*attr->data());
+}
+
+std::shared_ptr<AttrList> AttrList::parse(ieee1905_1::tlvWsc &tlv)
+{
+    if (!tlv.payload_length())
+        return nullptr;
+    std::shared_ptr<AttrList> attributes = nullptr;
+    //Check if its M1 or M2
+    auto msg_type = wsc_msg_type_from_buffer(tlv.payload(), tlv.payload_length());
+    if (msg_type == eWscMessageType::WSC_MSG_TYPE_M1) {
+        attributes = std::make_shared<m1>(tlv.payload(), tlv.payload_length(), true);
+        if (!attributes->init())
+            return nullptr;
+    } else if (msg_type == eWscMessageType::WSC_MSG_TYPE_M2) {
+        attributes = std::make_shared<m2>(tlv.payload(), tlv.payload_length(), true);
+        if (!attributes->init())
+            return nullptr;
```

Comment:
Move outside the if statement

Reply:
Also, you should make sure that the allocation was successful:
`if (!attributes || !attributes->init())`